File: ruby/example_code/dynamodb/GettingStarted/MoviesLoadData.rb
# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: Apache-2.0

# This code example demonstrates how to load information from a file into
# an existing table in Amazon DynamoDB named 'Movies'. The file must be named
# 'moviedata.json', must be in the same directory as this code example file,
# and must follow a specific format. For example, each table item for this
# particular table must be represented in the file with the following
# minimum information:
#
# {
#   "year": 2015,
#   "title": "The Big New Movie"
# }
#
# Table items can have additional information though, for example:
#
# [
#   {
#     "year": 2015,
#     "title": "The Big New Movie",
#     "info": {
#       "rating": 5.5,
#       "plot": "A big new movie launches.",
#       "genres": [
#         "Biography",
#         "Drama"
#       ],
#       "actors": [
#         "Larry",
#         "Moe",
#         "Curly"
#       ]
#     }
#   },
#   {
#     "year": 2017,
#     "title": "The Big New Movie 2",
#     "info": {
#       "rating": 3.0,
#       "plot": "Another big new movie launches.",
#       "running_time_secs": 7380
#       "actors": [
#         "Tina",
#         "Amy"
#       ]
#     }
#   }
# ]

# snippet-start:[dynamodb.Ruby.CodeExample.MoviesLoadData]
require "aws-sdk-dynamodb"
require "json"

$movie_counter = 0
$total_movies = 0

def add_item_to_table(dynamodb_client, table_item)
  dynamodb_client.put_item(table_item)
  $movie_counter += 1
  puts "Added movie #{$movie_counter}/#{$total_movies}: " \
    "'#{table_item[:item]['title']} " \
    "(#{table_item[:item]['year']})'."
rescue StandardError => e
  puts "Error adding movie '#{table_item[:item]['title']} " \
    "(#{table_item[:item]['year']})': #{e.message}"
  puts "Program stopped."
  exit 1
end

def run_me
# Replace us-west-2 with the AWS Region you're using for Amazon DynamoDB.
  region = "us-west-2"
  table_name = "Movies"
  data_file = "moviedata.json"

  # To use the downloadable version of Amazon DynamoDB,
  # uncomment the endpoint statement.
  Aws.config.update(
    # endpoint: 'http://localhost:8000',
    region: region
  )

  dynamodb_client = Aws::DynamoDB::Client.new
  file = File.read(data_file)
  movies = JSON.parse(file)
  $total_movies = movies.count

  puts "Adding #{$total_movies} movies from file '#{data_file}' " \
    "into table '#{table_name}'..."

  movies.each do |movie|
    table_item = {
      table_name: table_name,
      item: movie
    }
    add_item_to_table(dynamodb_client, table_item)
  end

  puts "Done."
end

run_me if $PROGRAM_NAME == __FILE__
# snippet-end:[dynamodb.Ruby.CodeExample.MoviesLoadData]

File: php/example_code/dynamodb/GettingStarted/MoviesLoadData.php
<?php
// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

// snippet-start:[dynamodb.php.codeexample.MoviesLoadData]
require 'vendor/autoload.php';

date_default_timezone_set('UTC');

use Aws\DynamoDb\Exception\DynamoDbException;
use Aws\DynamoDb\Marshaler;

$sdk = new Aws\Sdk([
    'endpoint'   => 'http://localhost:8000',
    'region'   => 'us-west-2',
    'version'  => 'latest'
]);

$dynamodb = $sdk->createDynamoDb();
$marshaler = new Marshaler();

$tableName = 'Movies';

$movies = json_decode(file_get_contents('moviedata.json'), true);

foreach ($movies as $movie) {
    $year = $movie['year'];
    $title = $movie['title'];
    $info = $movie['info'];

    $json = json_encode([
        'year' => $year,
        'title' => $title,
        'info' => $info
    ]);

    $params = [
        'TableName' => $tableName,
        'Item' => $marshaler->marshalJson($json)
    ];

    try {
        $result = $dynamodb->putItem($params);
        echo "Added movie: " . $movie['year'] . " " . $movie['title'] . "\n";
    } catch (DynamoDbException $e) {
        echo "Unable to add movie:\n";
        echo $e->getMessage() . "\n";
        break;
    }
}

// snippet-end:[dynamodb.php.codeexample.MoviesLoadData]
